sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -91,6 +91,16 @@
  */
 public abstract class DoFn<InputT extends @Nullable Object, OutputT extends @Nullable Object>
     implements Serializable, HasDisplayData {
+  /** Information accessible while within the {@link Setup} method. */
+  @SuppressWarnings("ClassCanBeStatic") // Converting class to static is an API change.
+  public abstract class SetupContext {
```

Comment:
You shouldn't need this. This is the "old" DoFn way of doing it.

Reply:
You mean we call invokeSetup() directly with PipelineOptions? I thought we always define a context for invoking the methods in DoFn, so we can add more parameters if needed in the future (like StartBundleContext, which only has pipelineOptions inside). If this not needed, I will go ahead and remove this context.

Reply:
I mean you would call `invokeSetup` with an `ArgumentProvider` that provides the options (this already exists). Then the job of the invoker (generated bytecode) is to pass the right arguments to the DoFn.

Basically: no more contexts passed to annotation-style `DoFn`s. In fact, we should deprecate the existing ones. Instead, `ArgumentProvider` is basically the same thing, but only passed to `DoFnInvoker` and the invoker passes just the arguments needed to the `DoFn`.

Reply:
Right, currently my impl is to pass this SetupContext to ArgumentProvider to get the arguments. I can remove this context and pass in the PipelineOptions directly.

Regarding to removing all contexts: I feel at least ProcessContext is kind of useful from the user point of view. Since @ProcessElement can take in quite a few parameters, it's a bit hard for the users to remember each kind of them (sometimes even I have to look it up to see what's the name of the types). ProcessContext does provide a good entry point to access element, output receiver, and side input.

Reply:
Makes sense. The explicit parameters in the args does make static analysis and validation possible. I agree there is no need to have a SetupContext. Updated the patch to remove it.